### ISSUE TITLE
Some bugs and doubts

### ISSUE THREAD

**Issue by ale94mleon** [Acellera/parameterize#32]
Hello,
I recently started to use parameterize. The thing is that until now I had gotten a lot of errors which I don't know how to handle. I am using parameterize in a computer cluster inside a conda environment which was created with the following commands:
```
conda create -n htmd
conda activate htmd
conda install psi4 psi4-rt python=3.7 -c psi4
conda install -c acellera -c psi4 -c conda-forge htmd
```
Because I use psi4 for another calculations too.
I launch the calculation in the cluster with the following bash script:
```
#!/bin/bash
#SBATCH --partition deflt #which partition I want
#SBATCH --output cis1_bh267_m_vacumm_Cell_882_SP_param_myjob.out #path for the slurm output
#SBATCH --error cis1_bh267_m_vacumm_Cell_882_SP_param_myjob.err #path for the slurm error output
#SBATCH --ntasks=3
#SBATCH --cpus-per-task=4 #number of cpu(logical cores)/task (task is normally an MPI process, default is one and the option to change it is -n)
#SBATCH --mem=14000 #amount of real memory per allocated CPU required by the job.
#SBATCH --time 2-00:00 #how many time I want the resources (this impacts the job priority as well)
#SBATCH --job-name=cis1_bh267_m_vacumm_Cell_882_SP_param #(to recognize your jobs when checking them with "squeue -u USERID")
#SBATCH --nodes=1 #number of node, usually 1 when no parallelization over nodes
#SBATCH --nice=0 #lowering your priority if > 0
#SBATCH --gpus=0 #number of gpu you want
#SBATCH --exclude=fang[1,11-50] #exclude some nodes
                                   
# This block is for the execution of the program
source /home/users/all-jh/opt/miniconda3/etc/profile.d/conda.sh #It is needed to use the conda activate command
conda activate htmd

# Creating local scratch folder for the user on the computing node.
MY_TEMP_DIR="$(mktemp -d /localdisk/psi4_${SLURM_JOBID}_$(date +%Y-%m-%d-%H-%M-%S)-XXXXXXXXXX)"
# make sure the temp dir was created
[ -d $MY_TEMP_DIR ] || exit 1
# make sure to delete the directory on (erroneous) exit
trap 'rm -rf -- "$MY_TEMP_DIR"' EXIT
# export PSI_SCRATCH env variable
export PSI_SCRATCH=$MY_TEMP_DIR

# This block is echoing some SLURM variables
echo "Job execution start: $(date)"
echo "JobID = $SLURM_JOBID"
echo "Host = $SLURM_JOB_NODELIST"
echo "Jobname = $SLURM_JOB_NAME"
echo "Subcwd = $SLURM_SUBMIT_DIR"
echo "SLURM_TASKS_PER_NODE = $SLURM_TASKS_PER_NODE"
echo "SLURM_CPUS_PER_TASK = $SLURM_CPUS_PER_TASK"
echo "SLURM_CPUS_ON_NODE = $SLURM_CPUS_ON_NODE"
echo "The temp file used was: $MY_TEMP_DIR"

#run paremeterize
parameterize ${SLURM_JOB_NAME}.mol2 --memory 14000 --ncpus 4 --theory HF --basis 6-31+G* --environment vacuum --min-type qm --charge-type ESP --scan-type qm --dihed-fit-type iterative --dihed_num_iterations 100 --outdir $SLURM_SUBMIT_DIR  
#Deleting the scratch file. Normal delete on normal exit
rm -rf $MY_TEMP_DIR

echo "Job execution end: $(date)"
```
If I try with B3LYP the quantum optimization calculation for the dihedral angles unexpectedly stop, so no data is available for the fitting processes (the folder parameters is not generated). Only with HF this is done. But the problem appears with the fitting process. These are the final lines of the error (Attached the entire _**2>stdout**_ file):

> Traceback (most recent call last):
>   File "/home/users/all-jh/opt/miniconda3/envs/htmd/bin/parameterize", line 11, in <module>
>     sys.exit(main_parameterize())
>   File "/home/users/all-jh/opt/miniconda3/envs/htmd/lib/python3.7/site-packages/parameterize/parameterization/cli.py", line 986, in main_parameterize
>     parameters = df.run()
>   File "/home/users/all-jh/opt/miniconda3/envs/htmd/lib/python3.7/site-packages/parameterize/parameterization/dihedral.py", line 621, in run
>     self._check()
>   File "/home/users/all-jh/opt/miniconda3/envs/htmd/lib/python3.7/site-packages/parameterize/parameterization/dihedral.py", line 615, in _check
>     assert np.isclose(self.loss, loss, rtol=0, atol=1e-2)
> AssertionError

[cis1_bh267_m_vacumm_Cell_882_SP_param_myjob.txt](https://github.com/Acellera/parameterize/files/5981082/cis1_bh267_m_vacumm_Cell_882_SP_param_myjob.txt)

The other issue is that if I try to reuse the qm calculation with the following command:
`parameterize cis1_bh267_m_vacumm_Cell_882_SP_param.mol2 --charge 1 --forcefield GAFF2 &> out.log &`
A lot of error came up. And my felling is that this command does not reuse the qm calculation but try to do them again but now with wB97X-D/6-311++G** (the file out.log is also attached).
[out.log](https://github.com/Acellera/parameterize/files/5981173/out.log)
Thank you so much for all your help in advance,
Alejandro

**Comment by raimis** [comment]
The argument of `parameterize` you are using would result into quite expensive and unreliable procedure. Unless, you know exactly what you are doing, I would recommend to use the default parameters. The defaults correspond to the procedure used in the paper (https://pubs.acs.org/doi/abs/10.1021/acs.jcim.9b00439) and proved to work variety of molecules. 

You just need to change `--memory` and `--ncpus` accordingly.

**Comment by ale94mleon** [comment]
Thank you so much for your replay.
If I understand correctly the expensive part of the calculation is the QM optimization. However this is not a problem for my chemical system and the available computational resource. Indeed, every single opt calculation was successfully done for the 36 conformation of each rotatable bond (in my case 6). The problem is not on this stage but in the fitting itself. The directories:
`minimize`
`esp`
`dihedral-opt`
`parameters`
are generated. But the directory `parameters` only has: parameters/GAFF2/`random-search.log`; nothing more. So, what is exactly done on this stage? Why could fail the fitting? Is this step computational expensive?
The error described above looks like the number of iteration is reached and the fitting was not possible, but is only a felling because I really don't know what is done here. If there is some bibliography that explain the math and the computational of this part I will appreciate. I will read again the paper that you sent me.  
I will also try with your advise and default values and send you the feedback. However for my system is highly recommended use even MP2 level of theory because the presence of the dihedral bond -CF2-CF3.
All the best and waiting for your replay,
Alejandro

**Comment by ale94mleon** [comment]
> The argument of `parameterize` you are using would result into quite expensive and unreliable procedure. Unless, you know exactly what you are doing, I would recommend to use the default parameters. The defaults correspond to the procedure used in the paper (https://pubs.acs.org/doi/abs/10.1021/acs.jcim.9b00439) and proved to work with variety of molecules.
> 
> You just need to change `--memory` and `--ncpus` accordingly.

**Comment by raimis** [comment]
A general problem that using QM for the dihedral scanning of a flexible molecule with several dihedrals results into broken bonds and or some strange geometries. As a results, the fitting algorithm cannot handle that. Note, the default behaviour is to use the initial FF for the scanning and QM is used just for energy calculations. 

> However for my system is highly recommended use even MP2 level of theory because the presence of the dihedral bond -CF2-CF3.

`parameterize` exposes just DFT, but you can easily hack the underlying Psi4 to run MP2 calculations.

**Comment by ale94mleon** [comment]
Dear Raimondas Galvelis,
Indeed you are totally right, I checked the structures and some of them have some really strange geometry (there are not many but exists) because the close contact between atoms in the initial structure that QM identifies as a bond and treated during the opt as such. Is not possible first check the structure looking for clash and discarded if they exist and/or optimize the structure at MM level with the fix dihedral and after that optimize at qm level? On this paper https://doi.org/10.1007/s00894-018-3911-5 they use "relaxed dihedral angle scans at the level of MP2/6-31G* and single point MP2/aug-cc-pVTZ calculations based on the optimized structures for assessing the energy profiles of the dihedral angle scan" (they are getting new GAFF2 parameters and use Gaussian 09). It looks like they don't get this "strange structure problem", maybe Gaussian has some protocol to get rid of this structures during the scan.
I also see on `parameterize/qm/custom.py` for the Class `class OMMMinimizer(Minimizer):` that the method `minimize` uses a deep potential (I presume) of -10000 kcal/mol to get fix the dihedral angle. However, in the aforementioned paper the authors say that: "a too strong potential causes very strong forces, which frequently cause the geometry optimization to crash, while a too weak constraining potential causes deviations from the desired dihedral angle. Therefore, the overall relaxation was carried out iteratively as a series of consecutive geometry optimizations with increasing constraining potential" How does parameterize solve this issue? 
Patch parameterize with MP2 is a really good idea. Psi4 is quite optimized to handle with MP2 because it uses by default the Density-Fitting (DF) approximation, which gives a really good efficiency to the quantum job (the factor "accuracy/time consuming" is pretty good) I have been looking through the master and I find the next lines that need to be changed, it would be of incredible help your opinion about it:
1. **`parameterize/parameterize/qm/base.py /`**
In:
`class QMBase(ABC, ProtocolInterface):`
Add to THEORIES (lines 65-75)
    THEORIES = (
        **"MP2",**
        "HF",
        "BLYP",
        "PBE",
        "B3LYP",
        "PBE0",
        "B2PLYP",
        "wB97",
        "wB97X",
        "wB97X-D",
    )

2. **`parameterize/parameterize/qm/psi4.py`**
In `class Psi4(QMBase):`
Line (140)
Change:
`reference += "hf" if self.theory == "HF" else "ks"`
by:
`reference += "hf" if self.theory in ["HF", "MP2"] else "ks"`
3. **`parameterize/parameterize/qm/gaussian.py `**
line 44: Change `if self.theory == "HF"` by `if self.theory in ["HF", "MP2"]`
line 127: `if "SCF Done:  E(RHF) = " in l:` I don't know how Gaussian print this line for MP2
4. **`parameterize/parameterize/qm/test.py`** 
A lot of things :-)
5. **`parameterize/parameterize/parameterization/cli.py`** 
line 68: Change `choices=["HF", "B3LYP", "wB97X-D"]` by `choices=["MP2", "HF", "B3LYP", "wB97X-D"]`
I have a doubt here because the amount of level of theory and basis set do not match with all the theories and basis set on **`parameterize/parameterize/qm/base.py /`**, why is that?

Finally, the directory `parameterize/parameterize/test-data/test-param/` would be of incredible help (as a tutorial or just consult) if each test had its respective used command.
I had not run any simulation on this days because the cluster of my Uni is really full, so, when I get some results I write again.
All the best and thank you so much for your help,
Alejandro

**Comment by ale94mleon** [comment]
I forgot,
This is, of course, only a suggestion: Until I know, Psi4 has scaling problem when the solvent is used (an issue in progress for the PSi4 team), so, if more than 1 CPU is desired for the qm calculation, vacuum is hardly recommended. In my experience is almost impossible to carry on the first stage of getting the charge on parameterize when the solvent PCM is included.
(https://github.com/psi4/psi4/issues/1280)

**Comment by ale94mleon** [comment]
I got really nice results :+1: 
MP2/aug-cc-pVTZ works really nice. My molecule has 24 atoms, and I optimize the six available rotable bonds.

1. The time of the total job was arond 14 hours.
2. The node used have the following characteristic: Xeon 6-core E-2136 @ 3.30GHz (12 logical cores) and Nvidia GTX 1070Ti.
3. The script used was the following: 
[cis1_bh267_m_vacumm_Cell_682_SP_param .sh.txt](https://github.com/Acellera/parameterize/files/6018497/cis1_bh267_m_vacumm_Cell_682_SP_param.sh.txt)

I used 12 CPU for parameterize because if not, the available memory on the node is not enough and the job get really slow because the rest ones need to way for the resource. In this way the jobs are only parallelized by Psi4 each at a time.
I hope that my problem and the solution could help to some one else.
One more to the collection:
![C-C0-C1-C2](https://user-images.githubusercontent.com/75276620/108638889-c50d5f80-7491-11eb-8bfa-6eda7f2e9ab0.png)
Happy simulations!
